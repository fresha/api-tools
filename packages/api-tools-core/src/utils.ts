import type { JSONValue, JSONObject } from './types';

export type KeyTransformFunc = (key: string) => string;
export type TransformFunc = (obj: JSONValue) => JSONValue;

export const identity: TransformFunc = obj => obj;

export const transformKeysDeep = (obj: JSONValue, keyFn: KeyTransformFunc): JSONValue => {
  if (Array.isArray(obj)) {
    return obj.map(o => transformKeysDeep(o, keyFn));
  }
  if (obj instanceof Object) {
    return Object.entries(obj).reduce((accum: JSONObject, [key, value]) => {
      const newKey = keyFn(key);
      return { ...accum, [newKey]: transformKeysDeep(value, keyFn) };
    }, {});
  }
  return obj;
};

const RE_WORD_START = /[A-Z][^A-Z]/g;

export const kebabCase: KeyTransformFunc = str => {
  const parts = str.replace(RE_WORD_START, r => `-${r.toLowerCase()}`).match(/[a-z0-9]+/g);
  return parts != null ? parts.filter(elem => elem.length).join('-') : str;
};

export const kebabCaseDeep: TransformFunc = obj => transformKeysDeep(obj, kebabCase);

const RE_WORD_SEPARATOR = /[-_]/g;

export const camelCase: KeyTransformFunc = str => {
  return str
    .split(RE_WORD_SEPARATOR)
    .filter(Boolean)
    .map((s, index) => (index > 0 ? s.slice(0, 1).toUpperCase() + s.slice(1) : s))
    .join('');
};

export const camelCaseDeep: TransformFunc = obj => transformKeysDeep(obj, camelCase);

export const titleCase = (str: string): string => {
  const camelized = camelCase(str);
  return `${camelized[0].toUpperCase()}${camelized.slice(1)}`;
};

export const snakeCase = (str: string): string => {
  return str
    .replace(/^[-_]+/, '')
    .replace(/^[A-Z]/, (match: string): string => match.toLowerCase())
    .replace(/-+/g, '_')
    .replace(/[-_]+[A-Z]/g, (match: string): string => String(match.at(-1)))
    .replace(
      /[a-z0-9][A-Z]/g,
      (match: string): string => `${String(match.at(0))}_${String(match.at(-1)).toLowerCase()}`,
    );
};
