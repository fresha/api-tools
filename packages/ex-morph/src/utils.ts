import fs from 'fs';

import { fs as memfs } from 'memfs';

export type FS = typeof fs | typeof memfs;

export const createFileSystem = (useInMemoryFileSystem: boolean): FS => {
  let result: FS = fs;
  if (useInMemoryFileSystem) {
    result = memfs;
    result.mkdirSync(process.cwd(), { recursive: true });
  }
  return result;
};

export const significantNameParts = (pathUrl: string): string[] =>
  pathUrl.split('/').filter(x => x && !(x.startsWith('{') && x.endsWith('}')));
