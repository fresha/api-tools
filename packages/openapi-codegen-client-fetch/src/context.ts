import type { Nullable } from '@fresha/api-tools-core';
import type { TSProjectContext } from '@fresha/openapi-codegen-utils';
import type { OperationModel } from '@fresha/openapi-model/build/3.0.3';
import type { SourceFile } from 'ts-morph';

export type NamingConvention = 'camel' | 'kebab' | 'snake' | 'title';

export interface Context extends TSProjectContext {
  readonly includeDeprecated: boolean;
  readonly includeInternal: boolean;
  readonly includedTags: Set<string>;
  readonly excludedTags: Set<string>;
  readonly apiNaming: Nullable<NamingConvention>;
  readonly clientNaming: Nullable<NamingConvention>;
}

export interface ActionContext extends Context {
  readonly operation: OperationModel;
  readonly sourceFile: SourceFile;
  readonly typesFile: SourceFile;
}
