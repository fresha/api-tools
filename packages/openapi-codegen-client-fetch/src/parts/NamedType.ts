import type { ActionContext } from '../context';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

export abstract class NamedType {
  readonly context: ActionContext;
  readonly name: string;
  readonly schema: SchemaModel;
  readonly isRequestBody: boolean;
  readonly isRequired: boolean;

  constructor(
    context: ActionContext,
    name: string,
    schema: SchemaModel,
    isRequestBody: boolean,
    isRequired: boolean,
  ) {
    this.context = context;
    this.name = name;
    this.schema = schema;
    this.isRequestBody = isRequestBody;
    this.isRequired = isRequired;
  }

  abstract collectData(namedTypes: Map<string, NamedType>): void;
  abstract generateCode(generatedTypes: Set<string>): void;
}
