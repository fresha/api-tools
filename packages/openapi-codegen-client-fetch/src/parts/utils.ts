import {
  camelCase,
  kebabCase,
  type JSONValue,
  type Nullable,
  snakeCase,
  titleCase,
} from '@fresha/api-tools-core';

import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

export type NamingConvention = 'camel' | 'kebab' | 'snake' | 'title';

export const propertyName = (name: string, convention: Nullable<NamingConvention>): string => {
  let result;
  switch (convention) {
    case 'camel':
      result = camelCase(name);
      break;
    case 'kebab':
      result = kebabCase(name);
      break;
    case 'snake':
      result = snakeCase(name);
      break;
    case 'title':
      result = titleCase(name);
      break;
    default:
      result = name;
  }
  return result.match(/^[_A-Za-z][_0-9A-Za-z]*$/) ? result : `'${result}'`;
};

export const objectPropertyName = (objectName: string, propName: string): string => {
  return propName.match(/^[_A-Za-z][_0-9A-Za-z]*$/)
    ? `${objectName}.${propName}`
    : `${objectName}[${propName}]`;
};

let i = 0;

export const schemaToType = (
  schema: Nullable<SchemaModel>,
  convention: Nullable<NamingConvention>,
): string => {
  switch (schema?.type) {
    case undefined:
    case null: {
      if (schema?.oneOfCount) {
        const elements: string[] = [];
        for (const subSchema of schema.oneOf()) {
          elements.push(schemaToType(subSchema, convention));
        }
        return elements.join(' | ');
      }

      return `Unknown${++i}`; // eslint-disable-line no-plusplus
    }
    case 'boolean':
      return schema?.nullable ? 'boolean | null' : 'boolean';
    case 'integer':
    case 'number': {
      const elements: JSONValue[] = [];
      if (schema.allowedValueCount) {
        elements.push(...schema.allowedValues());
      } else {
        elements.push('number');
      }
      if (schema.nullable) {
        elements.push('null');
      }
      return elements.join(' | ');
    }
    case 'string': {
      const elements: JSONValue[] = [];
      if (schema.allowedValueCount) {
        for (const val of schema.allowedValues()) {
          elements.push(`'${val as string}'`);
        }
      } else {
        elements.push('string');
      }
      if (schema.nullable) {
        elements.push('null');
      }
      return elements.join(' | ');
    }
    case 'array': {
      const subtype = schemaToType(schema.items, convention);
      return schema.nullable ? `${subtype}[] | null` : `${subtype}[]`;
    }
    case 'object': {
      const elements: string[] = [];
      for (const prop of schema.getProperties()) {
        elements.push(
          `${propertyName(prop.name, 'camel')}${prop.required ? '' : '?'}: ${schemaToType(
            prop.schema,
            convention,
          )}`,
        );
      }
      return schema.nullable ? `{ ${elements.join('; ')} } | null` : `{ ${elements.join('; ')} }`;
    }
    default:
      return `Unknown${++i}`; // eslint-disable-line no-plusplus
  }
};

export const determineSchemaName = (
  schema: SchemaModel,
  convention: Nullable<NamingConvention>,
): string => {
  let primaryDataSchemaName = schema.title;
  if (!primaryDataSchemaName && schema.root.components === schema.parent) {
    for (const [title, sharedSchema] of schema.root.components.schemas()) {
      if (schema === sharedSchema) {
        primaryDataSchemaName = title;
      }
    }
  }
  if (!primaryDataSchemaName) {
    primaryDataSchemaName = schemaToType(null, convention);
  }
  return primaryDataSchemaName;
};
