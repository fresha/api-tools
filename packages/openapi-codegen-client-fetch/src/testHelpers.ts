// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
