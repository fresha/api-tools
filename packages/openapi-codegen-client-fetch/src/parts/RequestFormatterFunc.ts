import assert from 'assert';

import { addFunction, addImportDeclaration } from '@fresha/code-morph-ts';
import {
  camelCase,
  getOperationIdOrThrow,
  RelationshipCardinality,
} from '@fresha/openapi-codegen-utils';
import { CodeBlockWriter, FunctionDeclaration, SyntaxKind } from 'ts-morph';

import { objectPropertyName, propertyName, schemaToType } from './utils';

import type { DocumentType } from './DocumentType';
import type { ActionContext } from '../context';
import type { OperationModel } from '@fresha/openapi-model/build/3.0.3';

const requestFormatterName = (operation: OperationModel) => {
  const operationId = getOperationIdOrThrow(operation);
  return camelCase(`format_${operationId}_request`);
};

export class RequestFormatterFunc {
  readonly context: ActionContext;
  readonly name: string;
  readonly resultType: DocumentType;

  constructor(resultType: DocumentType) {
    this.context = resultType.context;
    this.name = requestFormatterName(this.context.operation);
    this.resultType = resultType;
  }

  collectData(): void {
    this.context.logger.info(`Collecting data for request formatter ${this.name}`);
    // const primaryResource = this.resultType.primaryResourceType();
    // for (const prop of primaryResource.schema
    //   .getPropertyDeepOrThrow('attributes')
    //   .getProperties()) {
    //   global.console.log(prop.name, prop.schema.type);
    // }
  }

  generateCode(): void {
    this.context.logger.info(`Generating code for request formatter: ${this.name}`);

    if (this.resultType) {
      addImportDeclaration(this.context.formattersFile, './types', `t:${this.resultType.name}`);
    }

    const primaryResource = this.resultType.primaryResourceType();
    const funcHasArgs =
      primaryResource.attributesSchema?.hasPropertiesDeep() || !!primaryResource.relationships.size;

    const func = addFunction(
      this.context.formattersFile,
      this.name,
      funcHasArgs
        ? {
            params: '{}',
          }
        : {},
      this.resultType.name,
      true,
    );

    this.generateParameters(func);

    func.addStatements((writer: CodeBlockWriter) => {
      writer.write('return ');
      writer.inlineBlock(() => {
        writer.write("jsonapi: { version: '1.0' },");

        writer.writeLine('data: ');
        writer.inlineBlock(() => {
          writer.writeLine(`type: '${String(primaryResource.resourceType)}',`);
          // writer.writeLine(`id: '${String(primaryResource.resourceType)}',`);
          this.generateAttributes(writer);
          this.generateRelationships(writer);
        });

        if (this.resultType.includedResourceTypeCount) {
          writer.write('included: ');
          writer.inlineBlock(() => {});
        }
      });
    });
  }

  protected generateParameters(func: FunctionDeclaration): void {
    const primaryResource = this.resultType.primaryResourceType();

    const paramsType = func
      .getParameterOrThrow('params')
      .getTypeNodeOrThrow()
      .asKindOrThrow(SyntaxKind.TypeLiteral);

    if (primaryResource.attributesSchema) {
      for (const { name, schema } of primaryResource.attributesSchema.getPropertiesDeep()) {
        paramsType.addProperty({
          name: propertyName(name, this.context.clientNaming),
          type: schemaToType(schema, this.context.clientNaming),
          hasQuestionToken: !primaryResource.attributesSchema.isPropertyRequired(name),
        });
      }
    }

    if (primaryResource.relationships.size) {
      for (const [relName, relDef] of primaryResource.relationships) {
        let paramType: string;
        switch (relDef.cardinality) {
          case RelationshipCardinality.Many:
            paramType = 'string[]';
            break;
          case RelationshipCardinality.One:
            paramType = 'string';
            break;
          case RelationshipCardinality.ZeroOrOne:
            paramType = 'string | null';
            break;
          default:
            assert.fail(`Unsupported cardinality ${String(relDef.cardinality)}`);
        }

        paramsType.addProperty({
          name: propertyName(relName, this.context.clientNaming),
          type: paramType,
          hasQuestionToken: !relDef.required,
        });
      }
    }
  }

  protected generateAttributes(writer: CodeBlockWriter): void {
    const primaryResource = this.resultType.primaryResourceType();

    writer.writeLine('attributes: ');
    writer.inlineBlock(() => {
      // TODO need SchemaModel#propertyDeepCount property, returning a number of
      // properties in this schema and its allOf subschemas
      if (primaryResource.attributesSchema) {
        for (const { name } of primaryResource.attributesSchema.getPropertiesDeep()) {
          const propName = propertyName(name, this.context.clientNaming);
          writer.writeLine(`${propName}: ${objectPropertyName('params', propName)},`);
        }
      }
    });
    writer.write(',');
  }

  protected generateRelationships(writer: CodeBlockWriter): void {
    const primaryResource = this.resultType.primaryResourceType();

    writer.writeLine('relationships: ');
    writer.inlineBlock(() => {
      if (primaryResource.relationships.size) {
        for (const [relName, relDef] of primaryResource.relationships) {
          const paramName = propertyName(relName, this.context.clientNaming);
          const objPropName = objectPropertyName('params', paramName);

          let str = '';

          switch (relDef.cardinality) {
            case RelationshipCardinality.ZeroOrOne:
              str = `{ data: ${objPropName} == null ? { type: '${relDef.resourceType}', id: ${objPropName} } : null }`;
              break;
            case RelationshipCardinality.One:
              str = `{ data: { type: '${relDef.resourceType}', id: ${objPropName} } }`;
              break;
            case RelationshipCardinality.Many:
              str = `{ data: ${objPropName}.map(id => ({ type: '${relDef.resourceType}', id })) }`;
              break;
            default:
              assert.fail(`Unsupported relationship cardinality ${String(relDef.cardinality)}`);
          }

          writer.writeLine(
            relDef.required
              ? `${paramName}: ${str},`
              : `${paramName}: ${objPropName} !== undefined ? ${str} : undefined,`,
          );
        }
      }
    });
  }
}
