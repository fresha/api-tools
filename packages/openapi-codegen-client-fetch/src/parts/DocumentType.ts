import { Nullable } from '@fresha/api-tools-core';
import { addImportDeclaration, addTypeAlias } from '@fresha/code-morph-ts';
import { assert } from '@fresha/openapi-codegen-utils';
import { SyntaxKind } from 'ts-morph';

import { NamedType } from './NamedType';
import { ResourceType } from './ResourceType';
import { determineSchemaName, schemaToType } from './utils';

import type { ActionContext } from '../context';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

export class DocumentType extends NamedType {
  protected isRequestBody: boolean;
  protected primaryResourceTypes: NamedType[];
  protected primaryDataIsArray: boolean;
  protected includedResourceTypes: NamedType[];

  constructor(context: ActionContext, name: string, schema: SchemaModel, isRequestBody: boolean) {
    super(context, name, schema);
    this.isRequestBody = isRequestBody;
    this.primaryResourceTypes = [];
    this.primaryDataIsArray = false;
    this.includedResourceTypes = [];
  }

  collectData(namedTypes: Map<string, NamedType>): void {
    const primaryDataSchema = this.schema.getProperty('data');
    if (primaryDataSchema) {
      this.determinePrimaryDataTypes(primaryDataSchema, namedTypes);
    }
    const includedSchema = this.schema.getProperty('included');
    if (includedSchema) {
      this.determineIncludedDataTypes(includedSchema, namedTypes);
    }
  }

  private determinePrimaryDataTypes(
    primaryDataSchema: Nullable<SchemaModel>,
    namedTypes: Map<string, NamedType>,
  ): void {
    const primaryDataSchemas: SchemaModel[] = [];

    if (primaryDataSchema?.type === 'array') {
      if (Array.isArray(primaryDataSchema.items)) {
        primaryDataSchemas.push(...primaryDataSchema.items);
      } else if (primaryDataSchema.items) {
        primaryDataSchemas.push(primaryDataSchema.items);
      }
      this.primaryDataIsArray = true;
    } else if (primaryDataSchema) {
      primaryDataSchemas.push(primaryDataSchema);
    }

    assert(primaryDataSchema, `Cannot determine type of primary data`, this.context.operation);

    const primaryDataAlternatives = [];
    for (const primaryDataAlt of primaryDataSchemas) {
      primaryDataAlternatives.push(
        ...(primaryDataAlt.oneOf ?? []),
        ...(primaryDataAlt.anyOf ?? []),
      );
    }
    if (primaryDataAlternatives.length) {
      for (const subschema of primaryDataAlternatives) {
        this.determineSinglePrimaryDataType(subschema, namedTypes);
      }
    } else if (primaryDataSchemas.length) {
      for (const subschema of primaryDataSchemas) {
        this.determineSinglePrimaryDataType(subschema, namedTypes);
      }
    }
  }

  private determineSinglePrimaryDataType(
    primaryDataSchema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    const primaryDataSchemaName = determineSchemaName(primaryDataSchema);

    let existingNamedType = namedTypes.get(primaryDataSchemaName);
    assert(
      !existingNamedType || existingNamedType.schema === primaryDataSchema,
      `Name ${primaryDataSchemaName} have duplicate associated schemas`,
      this.context.operation,
    );

    if (!existingNamedType) {
      existingNamedType = new ResourceType(
        this.context,
        primaryDataSchemaName,
        primaryDataSchema,
        this.isRequestBody,
      );
      namedTypes.set(primaryDataSchemaName, existingNamedType);

      existingNamedType.collectData(namedTypes);
    }

    this.primaryResourceTypes.push(existingNamedType);
  }

  private determineIncludedDataTypes(
    includedSchema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    assert(
      includedSchema.type === 'array',
      `'included' field must be an array schema, but it is an ${String(includedSchema.type)}`,
      this.context.operation,
    );

    const includedItemsSchema = includedSchema.items;
    if (!includedItemsSchema) {
      return;
    }

    const includedItemsSchemas: SchemaModel[] = [];
    if (Array.isArray(includedItemsSchema)) {
      includedItemsSchemas.push(...includedItemsSchema);
    } else {
      includedItemsSchemas.push(includedItemsSchema);
    }

    for (const s of includedItemsSchemas) {
      if (s.type === 'object') {
        this.determineSingleIncludedDataType(s, namedTypes);
      } else {
        // multiple types of resources
        const subshemas = [...(s.oneOf ?? []), ...(s.anyOf ?? [])];

        for (const subschema of subshemas) {
          this.determineSingleIncludedDataType(subschema, namedTypes);
        }
      }
    }
  }

  private determineSingleIncludedDataType(
    schema: SchemaModel,
    namedTypes: Map<string, NamedType>,
  ): void {
    let schemaName: string;
    if (schema.parent === schema.root.components) {
      schemaName = schema.title ?? schemaToType(schema);
    } else {
      schemaName = schemaToType(schema);
    }

    let existingNamedType = namedTypes.get(schemaName);
    assert(
      !existingNamedType || existingNamedType.schema === schema,
      `Name ${schemaName} have duplicate associated schemas`,
      this.context.operation,
    );

    if (!existingNamedType) {
      existingNamedType = new ResourceType(this.context, schemaName, schema, this.isRequestBody);
      namedTypes.set(schemaName, existingNamedType);

      existingNamedType.collectData(namedTypes);
    }

    this.includedResourceTypes.push(existingNamedType);
  }

  generateCode(generatedTypes: Set<string>): void {
    if (!generatedTypes.has(this.name)) {
      generatedTypes.add(this.name);
    } else {
      return;
    }

    for (const resourceType of this.primaryResourceTypes) {
      resourceType.generateCode(generatedTypes);
    }
    for (const includedType of this.includedResourceTypes) {
      includedType.generateCode(generatedTypes);
    }

    const typeBaseName = this.isRequestBody ? 'JSONAPIClientDocument' : 'JSONAPIDataDocument';
    addImportDeclaration(this.context.typesFile, '@fresha/api-tools-core', `t:${typeBaseName}`);
    const typeAlias = addTypeAlias(this.context.typesFile, this.name, typeBaseName, true);

    const typeAliasArgs = [];
    if (this.primaryResourceTypes.length) {
      const typeName = this.primaryResourceTypes.map(t => t.name).join(' | ');
      typeAliasArgs.push(this.primaryDataIsArray ? `(${typeName})[]` : `${typeName}`);
    }
    if (this.includedResourceTypes.length) {
      if (!this.primaryResourceTypes.length) {
        typeAliasArgs.push(this.isRequestBody ? 'JSONAPIClientResource' : 'JSONAPIServerResource');
      }
      const names = this.includedResourceTypes.map(t => t.name).join(' | ');
      typeAliasArgs.push(`${names}`);
    }
    typeAlias
      .getNodeProperty('type')
      .asKindOrThrow(SyntaxKind.TypeReference)
      .addTypeArguments(typeAliasArgs);

    generatedTypes.delete(this.name);
  }
}
