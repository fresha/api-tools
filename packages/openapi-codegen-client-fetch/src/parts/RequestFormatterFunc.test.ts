import { Nullable } from '@fresha/api-tools-core';
import { buildEmployeeSchemasForTesting } from '@fresha/openapi-codegen-test-utils';
import {
  addResourceRelationships,
  MEDIA_TYPE_JSON_API,
  RelationshipCardinality,
  setDataDocumentSchema,
} from '@fresha/openapi-codegen-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { ActionContext } from '../context';
import { createActionTestContext } from '../testHelpers';

import { DocumentType } from './DocumentType';
import { NamedType } from './NamedType';
import { RequestFormatterFunc } from './RequestFormatterFunc';
import { NamingConvention } from './utils';

test.each([[null], ['camel'], ['snake'], ['kebab'], ['title']] as [Nullable<NamingConvention>][])(
  'client naming = %s',
  (clientNaming: Nullable<NamingConvention>) => {
    const openapi = OpenAPIFactory.create();

    buildEmployeeSchemasForTesting(openapi);

    const operation = openapi.setPathItem('/employees').addOperation('post');
    operation.operationId = 'createEmployee';

    const requestBodySchema = operation
      .setRequestBody()
      .setMediaType(MEDIA_TYPE_JSON_API)
      .setSchema('object');
    requestBodySchema.title = 'CreateEmployeeRequest';
    setDataDocumentSchema(requestBodySchema, 'employees');

    requestBodySchema
      .getPropertyDeepOrThrow('data')
      .getPropertyDeepOrThrow('attributes')
      .setProperties({
        'full-name': { type: 'string', required: true },
        age: { type: 'integer', required: true },
        gender: { type: 'string', enum: ['male', 'female'] },
        is_active: 'boolean',
      });

    addResourceRelationships(requestBodySchema.getPropertyDeepOrThrow('data'), {
      'Line-manager': {
        resourceType: 'employees',
        cardinality: RelationshipCardinality.One,
        required: true,
      },
      'subordinate-list': {
        resourceType: 'employees',
        cardinality: RelationshipCardinality.Many,
        required: false,
      },
      Mentor: {
        resourceType: 'employees',
        cardinality: RelationshipCardinality.ZeroOrOne,
        required: false,
      },
    });

    const namedTypes = new Map<string, NamedType>();
    const generatedTypes = new Set<string>();

    const context: ActionContext = {
      ...createActionTestContext(operation, 'src/index.ts'),
      clientNaming,
    };
    const requestType = new DocumentType(
      context,
      'CreateEmployeeRequest',
      requestBodySchema,
      true,
      true,
    );
    requestType.collectData(namedTypes);

    for (const namedType of namedTypes.values()) {
      namedType.generateCode(generatedTypes);
    }

    const requestFormatter = new RequestFormatterFunc(requestType);
    requestFormatter.collectData();
    requestFormatter.generateCode();

    expect(
      requestFormatter.context.project.getSourceFileOrThrow('src/formatters.ts').getText(),
    ).toMatchSnapshot('src/formatters.ts');
  },
);
