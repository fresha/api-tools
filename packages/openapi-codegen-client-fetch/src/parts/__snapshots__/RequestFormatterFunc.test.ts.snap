// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client naming = camel: src/formatters.ts 1`] = `
"import type { CreateEmployeeRequest } from "./types";

export function formatCreateEmployeeRequest(params: {
    fullName: string;
    age: number;
    gender?: 'male' | 'female';
    isActive?: boolean;
    lineManager: string;
    subordinateList?: string[];
    mentor?: string | null;
}): CreateEmployeeRequest {
    return {
        jsonapi: { version: '1.0' },
        data: 
        {
            type: 'employees',
            attributes: 
            {
                fullName: params.fullName,
                age: params.age,
                gender: params.gender,
                isActive: params.isActive,
            },
            relationships: 
            {
                lineManager: { data: { type: 'employees', id: params.lineManager } },
                subordinateList: params.subordinateList !== undefined ? { data: params.subordinateList.map(id => ({ type: 'employees', id })) } : undefined,
                mentor: params.mentor !== undefined ? { data: params.mentor == null ? { type: 'employees', id: params.mentor } : null } : undefined,
            }
        }
    }
}
"
`;

exports[`client naming = kebab: src/formatters.ts 1`] = `
"import type { CreateEmployeeRequest } from "./types";

export function formatCreateEmployeeRequest(params: {
    'full-name': string;
    age: number;
    gender?: 'male' | 'female';
    'is-active'?: boolean;
    'line-manager': string;
    'subordinate-list'?: string[];
    mentor?: string | null;
}): CreateEmployeeRequest {
    return {
        jsonapi: { version: '1.0' },
        data: 
        {
            type: 'employees',
            attributes: 
            {
                'full-name': params['full-name'],
                age: params.age,
                gender: params.gender,
                'is-active': params['is-active'],
            },
            relationships: 
            {
                'line-manager': { data: { type: 'employees', id: params['line-manager'] } },
                'subordinate-list': params['subordinate-list'] !== undefined ? { data: params['subordinate-list'].map(id => ({ type: 'employees', id })) } : undefined,
                mentor: params.mentor !== undefined ? { data: params.mentor == null ? { type: 'employees', id: params.mentor } : null } : undefined,
            }
        }
    }
}
"
`;

exports[`client naming = null: src/formatters.ts 1`] = `
"import type { CreateEmployeeRequest } from "./types";

export function formatCreateEmployeeRequest(params: {
    'full-name': string;
    age: number;
    gender?: 'male' | 'female';
    is_active?: boolean;
    'Line-manager': string;
    'subordinate-list'?: string[];
    Mentor?: string | null;
}): CreateEmployeeRequest {
    return {
        jsonapi: { version: '1.0' },
        data: 
        {
            type: 'employees',
            attributes: 
            {
                'full-name': params['full-name'],
                age: params.age,
                gender: params.gender,
                is_active: params.is_active,
            },
            relationships: 
            {
                'Line-manager': { data: { type: 'employees', id: params['Line-manager'] } },
                'subordinate-list': params['subordinate-list'] !== undefined ? { data: params['subordinate-list'].map(id => ({ type: 'employees', id })) } : undefined,
                Mentor: params.Mentor !== undefined ? { data: params.Mentor == null ? { type: 'employees', id: params.Mentor } : null } : undefined,
            }
        }
    }
}
"
`;

exports[`client naming = snake: src/formatters.ts 1`] = `
"import type { CreateEmployeeRequest } from "./types";

export function formatCreateEmployeeRequest(params: {
    full_name: string;
    age: number;
    gender?: 'male' | 'female';
    is_active?: boolean;
    line_manager: string;
    subordinate_list?: string[];
    mentor?: string | null;
}): CreateEmployeeRequest {
    return {
        jsonapi: { version: '1.0' },
        data: 
        {
            type: 'employees',
            attributes: 
            {
                full_name: params.full_name,
                age: params.age,
                gender: params.gender,
                is_active: params.is_active,
            },
            relationships: 
            {
                line_manager: { data: { type: 'employees', id: params.line_manager } },
                subordinate_list: params.subordinate_list !== undefined ? { data: params.subordinate_list.map(id => ({ type: 'employees', id })) } : undefined,
                mentor: params.mentor !== undefined ? { data: params.mentor == null ? { type: 'employees', id: params.mentor } : null } : undefined,
            }
        }
    }
}
"
`;

exports[`client naming = title: src/formatters.ts 1`] = `
"import type { CreateEmployeeRequest } from "./types";

export function formatCreateEmployeeRequest(params: {
    FullName: string;
    Age: number;
    Gender?: 'male' | 'female';
    IsActive?: boolean;
    LineManager: string;
    SubordinateList?: string[];
    Mentor?: string | null;
}): CreateEmployeeRequest {
    return {
        jsonapi: { version: '1.0' },
        data: 
        {
            type: 'employees',
            attributes: 
            {
                FullName: params.FullName,
                Age: params.Age,
                Gender: params.Gender,
                IsActive: params.IsActive,
            },
            relationships: 
            {
                LineManager: { data: { type: 'employees', id: params.LineManager } },
                SubordinateList: params.SubordinateList !== undefined ? { data: params.SubordinateList.map(id => ({ type: 'employees', id })) } : undefined,
                Mentor: params.Mentor !== undefined ? { data: params.Mentor == null ? { type: 'employees', id: params.Mentor } : null } : undefined,
            }
        }
    }
}
"
`;
